cpp/src/groupby/sort/group_tdigest.cu

```diff
@@ -0,0 +1,849 @@
+    mean_col.begin<double>(), weight_col.begin<double>(), thrust::make_discard_iterator()));
+  auto const num_values = std::distance(centroids_begin, centroids_end);
+  thrust::reduce_by_key(rmm::exec_policy(stream),
+                        keys,
+                        keys + num_values,                // keys
+                        centroids_begin,                  // values
+                        thrust::make_discard_iterator(),  // key output
+                        output,                           // output
+                        thrust::equal_to<size_type>{},    // key equality check
+                        merge_centroids{});
+
+  // create the list
+  auto const num_groups = group_cluster_offsets->size() - 1;
+  auto list             = cudf::make_lists_column(
+    num_groups, std::move(group_cluster_offsets), std::move(tdigests), 0, {});
+
+  // create final tdigest column
+  std::vector<std::unique_ptr<column>> children;
+  children.push_back(std::move(list));
+  children.push_back(std::move(min_col));
+  children.push_back(std::move(max_col));
+  return make_structs_column(num_groups, std::move(children), 0, {}, stream, mr);
+}
+
+// retrieve total weight of scalar inputs by group index
+struct scalar_total_weight {
+  size_type const* group_valid_counts;
+  __device__ double operator()(size_type group_index) { return group_valid_counts[group_index]; }
+};
+
+// return the min/max value of scalar inputs by group index
+template <typename T>
+struct get_scalar_minmax {
+  column_device_view const col;
+  device_span<size_type const> group_offsets;
+  size_type const* group_valid_counts;
+
+  __device__ thrust::tuple<double, double> operator()(size_type group_index)
+  {
+    // note: .element<T>() is taking care of fixed-point conversions for us.
+    return {static_cast<double>(col.element<T>(group_offsets[group_index])),
+            static_cast<double>(
+              col.element<T>(group_offsets[group_index] + (group_valid_counts[group_index] - 1)))};
+  }
+};
+
+struct typed_group_tdigest {
+  template <
+    typename T,
+    typename std::enable_if_t<cudf::is_numeric<T>() || cudf::is_fixed_point<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& col,
+                                     cudf::device_span<size_type const> group_offsets,
+                                     cudf::device_span<size_type const> group_labels,
+                                     cudf::device_span<size_type const> group_valid_counts,
+                                     size_type num_groups,
+                                     int delta,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+  {
+    // first, generate cluster weight information for each input group
+    auto total_weight = cudf::detail::make_counting_transform_iterator(
+      0, scalar_total_weight{group_valid_counts.begin()});
```

Comment:
I'm confused by this. This just looks like iterating over `group_valid_counts` but with extra steps. It seems like all this does is `*(total_weight + i) == group_valid_counts[i]`. 

Am I missing something?

Reply:
That's pretty much it.  In the scalar case, the # of (non-null) input values is the total weight.  I just wanted to give it an explicit name so it didn't get lost in the sea of parameters.

Reply:
How about just
```
auto total_weight = group_valid_counts.begin();
```
?